zkevm-circuits/src/evm_circuit/util/common_gadget.rs

```diff
@@ -89,29 +95,136 @@ impl<F: Field> SameContextGadget<F> {
 
         Self {
             opcode,
+            is_first_bytecode_table,
             sufficient_gas_left,
         }
     }
 
+    // Check if current bytecode is belong to first bytecode table.
+    pub(crate) fn is_first_sub_bytecode(&self) -> Expression<F> {
+        self.is_first_bytecode_table.expr()
+    }
+
     pub(crate) fn assign_exec_step(
         &self,
         region: &mut CachedRegion<'_, '_, F>,
         offset: usize,
-        _block: &Block,
-        _call: &Call,
+        block: &Block,
+        call: &Call,
         step: &ExecStep,
     ) -> Result<(), Error> {
         let opcode = step.opcode.unwrap();
+        let is_first_bytecode_table = block.is_first_sub_bytecode_circuit(&call.code_hash);
+
         self.opcode
             .assign(region, offset, Value::known(F::from(opcode.as_u64())))?;
 
+        self.is_first_bytecode_table.assign(
+            region,
+            offset,
+            Value::known(F::from(is_first_bytecode_table as u64)),
+        )?;
+
         self.sufficient_gas_left
             .assign(region, offset, F::from(step.gas_left - step.gas_cost))?;
 
         Ok(())
     }
 }
 
+#[derive(Clone, Debug)]
+pub(crate) struct BytecodeLengthGadget<F> {
+    pub(crate) code_length: Cell<F>,
+    // indicates current op code belongs to first or second bytecode table.
+    // should be bool type.
+    pub(crate) is_first_bytecode_table: Cell<F>,
+}
+
+impl<F: Field> BytecodeLengthGadget<F> {
+    pub(crate) fn construct(cb: &mut EVMConstraintBuilder<F>, code_hash: Cell<F>) -> Self {
+        // Fetch the bytecode length from the bytecode table.
+        let code_length = cb.query_cell();
+        let is_first_bytecode_table = cb.query_bool();
+
+        cb.condition(is_first_bytecode_table.clone().expr(), |cb| {
```

Comment:
add a comment to explain that if dual bytecode circuit not enabled, `is_first_bytecode_table` is always true?